subprojects/core/src/main/java/org/gradle/process/internal/DefaultSystemPropertyHandler.java

```diff
@@ -0,0 +1,180 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.process.internal;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.Named;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputDirectory;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.process.CommandLineArgumentProvider;
+import org.gradle.process.InputOutputValue;
+import org.gradle.process.JavaForkOptions;
+import org.gradle.process.SystemPropertyHandler;
+import org.gradle.util.DeferredUtil;
+
+public class DefaultSystemPropertyHandler implements SystemPropertyHandler {
+    private final JavaForkOptions javaForkOptions;
+
+    public DefaultSystemPropertyHandler(JavaForkOptions javaForkOptions) {
+        this.javaForkOptions = javaForkOptions;
+    }
+
+    @Override
+    public void add(String name, InputOutputValue value) {
+        javaForkOptions.getJvmArgumentProviders().add(new SystemPropertyCommandLineArgument(name, value));
+    }
+
+    @Override
+    public void add(String name, Object value) {
+        javaForkOptions.systemProperty(name, value);
+    }
+
+    @Override
+    public InputOutputValue inputFile(Object file) {
+        return new InputFileValue(file);
+    }
+
+    @Override
+    public InputOutputValue inputDirectory(Object file) {
+        return new InputDirectoryValue(file);
+    }
+
+    @Override
+    public InputOutputValue outputFile(Object file) {
+        return new OutputFileValue(file);
+    }
+
+    @Override
+    public InputOutputValue outputDirectory(Object file) {
+        return new OutputDirectoryValue(file);
+    }
+
+    @Override
+    public InputOutputValue ignored(Object file) {
+        return new IgnoredValue(file);
+    }
+
+    private class SystemPropertyCommandLineArgument implements CommandLineArgumentProvider, Named {
+        private final String name;
+        private final InputOutputValue inputOutputValue;
+
+        public SystemPropertyCommandLineArgument(String name, InputOutputValue inputOutputValue) {
+            this.name = name;
+            this.inputOutputValue = inputOutputValue;
+        }
+
+        @Nested
+        public InputOutputValue getValue() {
+            return inputOutputValue;
+        }
+
+        @Input
+        @Override
+        public String getName() {
+            return name;
+        }
```

Comment:
Should this be marked as `@Input` if it's already included in the name of the property?

Reply:
I can mark it as `@Internal`.

Reply:
After thinking a little bit more, I think that the `name` should be an input, given that it influences the outcome of `asArguments`.